### PR TITLE
Fix pyright errors in `arcade.gl`

### PR DESCRIPTION
**Pull request by cspotcode**
Followup to #1751.  Will need a rebase after #1751 gets squashed.  Requesting a review from @einarf , but no rush, I know you're slowly getting back into things.

### CODE REVIEW

File: arcade/cache/texture.py
Comment (gran4) [inline arcade/cache/texture.py]: Shouldn't the type be str and not Union[str, Path] if you are correct?

File: arcade/gl/context.py
Comment (gran4) [inline arcade/gl/context.py]: delete import typing

File: arcade/gl/program.py
Comment (gran4) [inline arcade/gl/program.py]: Shouldn't this be List[Tuple[str, int]] or does it matter?

File: arcade/gl/types.py
Comment (gran4) [inline arcade/gl/types.py]: List not list, again, does it matter?

File: arcade/gui/examples/scroll_area.py
Comment (gran4) [inline arcade/gui/examples/scroll_area.py]: I think you can remove [float], and stuff. Unless there is something I am misunderstanding

File: arcade/gui/property.py
Comment (gran4) [inline arcade/gui/property.py]: I remember doing this file, we might have some conflicts

File: arcade/gui/property.py
Comment (gran4) [inline arcade/gui/property.py]: Make it a Union[P, Property]?

File: arcade/gui/widgets/layout.py
Comment (gran4) [inline arcade/gui/widgets/layout.py]: List[List[Tuple[int, int]]] instead of list[list[tuple[int, int]]]

File: arcade/gui/widgets/slider.py
Comment (gran4) [inline arcade/gui/widgets/slider.py]: value_x looks like a float, so instead of making self.y + self.height // 2 an int, change -> Tuple[int, int]: to -> Tuple[float, float]:

File: arcade/texture/transforms.py
Comment (gran4) [inline arcade/texture/transforms.py]: Dict[Tuple[]]

**Comment by cspotcode** [comment]
@gran4 Thanks for taking a look, but unfortunately, this PR can't be reviewed yet.  It's stacked on top of #1751 so the diff is all wonky.  It'll be reviewable once #1751 merges.

**Comment by gran4** [comment]
Oh, ok you should make it a draft then. (unless there is something I don't understand)

File: arcade/cache/texture.py
Comment (einarf) [reply arcade/cache/texture.py]: I will redo all the cache stuff anyway, but this doesn't hurt.

**Comment by einarf** [comment]
The gl stuff looks perfectly reasonable 👍

**Review by einarf — APPROVED** [review]

**Comment by cspotcode** [comment]
@einarf Thanks for taking a look!  I marked this PR as draft while we wait for #1751 to merge.  Once #1751 is merged, I'll rebase this PR, bring it out of draft, and ping you for a merge.

File: arcade/gl/context.py
Comment (cspotcode) [reply arcade/gl/context.py]: It was actually used in this file, `typing.cast()`.  I did this because `cast` already refers to something else.  But I think I should remove that bit anyway.

**Comment by cspotcode** [comment]
I rebased yesterday, but there are still one or two mypy errors I need to understand and fix before this can merge.  I'll solicit a review after I've fixed them.

**Comment by cspotcode** [comment]
@einarf I need some help understanding some of the gl types.

https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/vertex_array.py#L195-L196
I get a typecheck diagnostic `Argument 1 to "gl_name" has incompatible type "Optional[int]"; expected "int"`
This is because `prog_attr.gl_type` is `Optional[int]`

Is that correct, should `AttribFormat.gl_type` be `Optional`?

`AttribFormat` is instantiated here:
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/program.py#L394-L401
and here:
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/types.py#L311-L314

For the latter, `gl_type` is possibly `None` because it comes from `_formats`, which has some `None` values in x1, x2, x4, x8
https://github.com/pythonarcade/arcade/blob/889772df9b9c94d77bbb09a500e9dcb0be180e53/arcade/gl/types.py#L196-L220

File: arcade/gl/types.py
Comment (cspotcode) [reply arcade/gl/types.py]: Good catch, python 3.8 needs it capitalized.

**Pull request merged**
